giz-berlin/nextcloud-oidc-login
# Add WebDAV access using providers bearer token

This PR adds the ability to allow WebDAV access using a providers bearer token. This can be useful when using [rclone WebDAV with OIDC](https://rclone.org/webdav/#openid-connect) to synchronize a users files from Nextcloud.

To avoid too much load on the OIDC provider, this PR also adds caching of the JWKs as well as the discovered OIDC configuration for a configurable amount of time.

## Testing

You can use the following script to test this functionality. To pass the tokens `audience` check in the `OpenIDConnectClient.php`, you will need to configure an audience mapper in Keycloak to ensure that your client is included in the `aud` property of the JWT. A [hardcoded audience mapper](https://www.keycloak.org/docs/latest/server_admin/#_audience_hardcoded) is sufficient here. Basically follow the following steps:

1. Go to `Client Scopes`
2. Add new client scope, call it `nextcloud`.
3. Under `Mappers` create a new mapper of type `Audience` and ensure that `Included Client Audience` contains your Nextcloud client. Click Save.
4. Finally, go to `Client > your-nextcloud-client > Client Scopes` and add the new `nextcloud` scope.

```bash
#!/bin/bash

KC_BASE="<keycloak-base>" # e.g. http://localhost:8080/auth
KC_USERNAME="<keycloak-username>"
KC_PASSWORD="<keycloak-password>"
KC_CLIENT="<your-nextcloud-client>"
KC_SECRET="<your-nextcloud-secret>"

NC_BASE="<nextcloud-base>" # e.g. http://localhost:8081
NC_UID="<nextcloud-user-id>"

resp=$(curl -s -X POST "$KC_BASE/realms/giz/protocol/openid-connect/token" \
 -H "Content-Type: application/x-www-form-urlencoded" \
 -d 'grant_type=password' \
 -d "client_id=$KC_CLIENT" -d "client_secret=$KC_SECRET" \
 -d "username=$KC_USERNAME" -d "password=$KC_PASSWORD" \
 -d "scope=nextcloud")

token=$(echo $resp | jq -r .access_token)

if [ ! -n "$token" ]; then
    echo $resp
    exit 1
fi

curl -X PROPFIND \
    -H "Depth: 1" \
    -H "Authorization: Bearer $token" \
    $NC_BASE/remote.php/dav/files/$NC_UID/
```
